code/modules/mafia/controller.dm

```diff
@@ -0,0 +1,578 @@
+		on_trial.body.forceMove(get_turf(on_trial.assigned_landmark))
+	//by now clowns should have killed someone in guilty list, clear this out
+	judgement_innocent_votes = list()
+	judgement_guilty_votes = list()
+	on_trial = null
+	//day votes are already cleared, so this will skip the trial and check victory/lockdown/whatever else
+	next_phase_timer = addtimer(CALLBACK(src, .proc/check_trial, FALSE),judgement_lynch_period,TIMER_STOPPABLE)// small pause to see the guy dead, no verbosity since we already did this
+
+/datum/mafia_controller/proc/send_home(datum/mafia_role/role)
+	role.body.forceMove(get_turf(role.assigned_landmark))
+
+/datum/mafia_controller/proc/check_victory()
+	var/alive_town = 0
+	var/alive_mafia = 0
+	var/list/solos_to_ask = list() //need to ask after because first round is counting team sizes
+	var/list/total_victors = null //if this has someone, they won alone. list because side antags can with with people
+	var/blocked_victory = FALSE
+
+	///PHASE ONE: TALLY UP ALL NUMBERS OF PEOPLE STILL ALIVE
+
+	for(var/datum/mafia_role/R in all_roles)
+		if(R.game_status == MAFIA_ALIVE)
+			switch(R.team)
+				if(MAFIA_TEAM_MAFIA)
+					alive_mafia++
+				if(MAFIA_TEAM_TOWN)
+					alive_town++
+				if(MAFIA_TEAM_SOLO)
+					if(R.solo_counts_as_town)
+						alive_town++
+					solos_to_ask += R
+
+	///PHASE TWO: SEND STATS TO SOLO ANTAGS, SEE IF THEY WON OR TEAMS CANNOT WIN
+
+	for(var/datum/mafia_role/solo in solos_to_ask)
+		if(solo.check_total_victory(alive_town, alive_mafia))
+			total_victors += solo
+		else if(solo.block_team_victory(alive_town, alive_mafia))
+			blocked_victory = TRUE
+
+	///PHASE THREE: IF SOLOS WON, SEND A SIGNAL THAT GAME IS ENDING (literally just for fugitives to say they won sorry not sorry)
+	if(length(total_victors))
+		SEND_SIGNAL(src,COMSIG_MAFIA_GAME_END)
+	//solo victories!
+	var/solo_end = FALSE
+	for(var/datum/mafia_role/winner in total_victors)
+		send_message("<span class='big red'>!! [uppertext(winner.name)] VICTORY !!</span>")
+		solo_end = TRUE
+	if(solo_end)
+		start_the_end()
+	if(alive_mafia == 0 && !blocked_victory)
+		SEND_SIGNAL(src,COMSIG_MAFIA_GAME_END)
+		start_the_end("<span class='big green'>!! TOWN VICTORY !!</span>")
+		return TRUE
+	else if(alive_mafia >= alive_town && !blocked_victory) //guess could change if town nightkill is added
+		SEND_SIGNAL(src,COMSIG_MAFIA_GAME_END)
+		start_the_end("<span class='big red'>!! MAFIA VICTORY !!</span>")
+		return TRUE
+
+/datum/mafia_controller/proc/start_the_end(message)
+	if(message)
+		send_message(message)
+	for(var/datum/mafia_role/R in all_roles)
+		R.reveal_role(src)
+	phase = MAFIA_PHASE_VICTORY_LAP
+	next_phase_timer = addtimer(CALLBACK(src,.proc/end_game),victory_lap_period,TIMER_STOPPABLE)
+
+/datum/mafia_controller/proc/end_game()
+
+	map_deleter.generate() //remove the map, it will be loaded at the start of the next one
```

Comment:
Using map generator here is bit pointless, just retrieve map template turfs and call empty on them